src/logic/otp2/controller.ts

```diff
@@ -536,15 +548,153 @@ async function searchTaxiFrontBack(
     }
 }
 
-export type NonTransitMode = 'foot' | 'bicycle' | 'car' | 'bike_rental'
+/**
+ * TODO 7/12-21: Clean up function signature and types once OTP1 is gone.
+ * As we have removed recursion, initialSearchDate can be replaced with
+ * searchDate. searchFilter could
+ * be kept separate from Otp2SearchParams.
+ */
+export async function searchTransit(
+    { initialSearchDate, searchFilter, ...searchParams }: Otp2SearchParams,
+    extraHeaders: { [key: string]: string },
+): Promise<TransitTripPatterns> {
+    const { searchDate, arriveBy } = searchParams
+
+    const getTripPatternsParams = {
+        ...searchParams,
+        modes: filterModesAndSubModes(searchFilter),
+    }
+
+    // initial search date may differ from search date if this is a
+    // continuation search using a cursor.
+    const isFirstSearchIteration = initialSearchDate === searchDate
+
+    // We do two searches in parallel here to speed things up a bit. One is a
+    // flexible search where we explicitly look for trips that may include means
+    // of transport that has to be booked in advance the second is a regular
+    // search.
+    const [
+        flexibleResults,
+        [regularTripPatternsUnfiltered, initialMetadata, routingErrors],
+    ] = await Promise.all([
+        isFirstSearchIteration ? searchFlexible(searchParams) : undefined,
+        getTripPatterns(getTripPatternsParams),
+    ])
+
+    let queries = [
+        ...(flexibleResults?.queries || []),
+        getTripPatternsQuery(getTripPatternsParams, 'First regular search'),
+    ]
+
+    let metadata = initialMetadata
+
+    let regularTripPatterns = regularTripPatternsUnfiltered
```

Comment:
Do `queries`, `metadata` and `regularTripPatterns` need to be mutated?

Reply:
I would love for them not to be! queries collects all performed queries so we must either store the intermediate queries separately and join them before returning, or mutate query. I think it happened even more places earlier but that's no excuse for trying to get rid of it completely.

regularTripPatterns DO change, or rather, which ones we return. It's the same here I guess, it all depends on how we store the intermediates and how we join them in the end. Please please help me find a better pattern!

Reply:
I've decided to do this in a later PR. I've tried to map out how a search actually works and I think we can improve things a bit, but this PR is big enough.